src/sick_mrs1000_parser.cpp

```diff
@@ -0,0 +1,339 @@
+  scan.scan_time = 1.0 / (scanning_freq / 100.0);
+  // ROS_DEBUG("hex: %s, scanning_freq: %d, scan_time: %f", fields[16], scanning_freq, scan.scan_time);
+
+  // 17: Measurement Frequency (36)
+  unsigned short measurement_freq = -1;
+  sscanf(fields[17], "%hx", &measurement_freq);
+  scan.time_increment = 1.0 / (measurement_freq * 100.0);
+  if (override_time_increment_ > 0.0)
+  {
+    // Some lasers may report incorrect measurement frequency
+    scan.time_increment = override_time_increment_;
+  }
+
+  // 23: Starting angle (FFF92230)
+  int starting_angle = -1;
+  sscanf(fields[23], "%x", &starting_angle);
+  scan.angle_min = (starting_angle / 10000.0) / 180.0 * M_PI - M_PI / 2;
+  // ROS_DEBUG("starting_angle: %d, angle_min: %f", starting_angle, scan.angle_min);
+
+  // 24: Angular step width (2710)
+  unsigned short angular_step_width = -1;
+  sscanf(fields[24], "%hx", &angular_step_width);
+  scan.angle_increment = (angular_step_width / 10000.0) / 180.0 * M_PI;
+  scan.angle_max = scan.angle_min + (number_of_data - 1) * scan.angle_increment;
+
+  // 25: Number of data (<= 10F)
+  // This is already determined above in number_of_data
+
+  // adjust angle_min to min_ang config param
+  int index_min = 0;
+  while (scan.angle_min + scan.angle_increment < current_config_.min_ang)
+  {
+    scan.angle_min += scan.angle_increment;
+    index_min++;
+  }
+
+  // adjust angle_max to max_ang config param
+  int index_max = number_of_data - 1;
+  while (scan.angle_max - scan.angle_increment > current_config_.max_ang)
+  {
+    scan.angle_max -= scan.angle_increment;
+    index_max--;
+  }
+
+  ROS_DEBUG("index_min: %d, index_max: %d", index_min, index_max);
+  // ROS_DEBUG("angular_step_width: %d, angle_increment: %f, angle_max: %f", angular_step_width, scan.angle_increment, scan.angle_max);
+
+  double phi = scan.angle_min;
+  double alpha = (-layer/ 100) * M_PI / 180;
+
+  // first layer
+  // order of layers: 0, -250, 250, -500
+  if(layer_count_ == 0){
+    modifier_.resize(4 * (index_max - index_min + 1));
+    point_counter_ = 0;
+    x_iter = sensor_msgs::PointCloud2Iterator<float>(cloud_, "x");
+    y_iter = sensor_msgs::PointCloud2Iterator<float>(cloud_, "y");
+    z_iter = sensor_msgs::PointCloud2Iterator<float>(cloud_, "z");
+  }
+  layer_count_++;
+
+  // 26..26 + n - 1: Data_1 .. Data_n
+  scan.ranges.resize(index_max - index_min + 1);
+  for (int j = index_min; j <= index_max; ++j)
+  {
+    unsigned short range;
+    sscanf(fields[j + 26], "%hx", &range);
+    float range_meter = range / 1000.0;
+    scan.ranges[j - index_min] = range_meter;
+
+    // Transform point from spherical coordinates to Cartesian coordinates.
+    *x_iter = range_meter * sin(M_PI_2 - alpha) * cos(phi);
+    *y_iter = range_meter * sin(M_PI_2 - alpha) * sin(phi);
+    *z_iter = range_meter * cos(M_PI_2 - alpha);
+
+    ++x_iter;
+    ++y_iter;
+    ++z_iter;
+
+    phi += scan.angle_increment;
+    point_counter_++;
+  }
+
+  if(layer_count_ == 4){
```

Comment:
I don't agree here, even with the heads up about measuring all 4 lines at once (thanks for clearing that up), this results in any 4 lines in the point cloud. I would expect it to rather be after receiving the 4 layer (=-500?). Maybe we should even check if they are all from the same swipe (telegram or scan counter maybe?).

Reply:
Hmmm since the packages are sent via tcp they are in the right order and there are no un-transfered packages, or? The layers are always in the right oder here. Or do you mean we should check for the first layer and then start counting? -- No idea.

Reply:
Start counting on the first layer should be ok (or only send on the fourth layer).

Reply:
ok.. will do that tomorrow. ;)

Reply:
Yes, it's necessary that we group all 4 scans that were recorded simultaneously into one point cloud, otherwise the time stamp will be wrong.